src/components/restore/RestoreIcloudStep.js

```diff
@@ -0,0 +1,299 @@
+    size={52}
+    start={{ x: 1, y: 1 }}
+    steps={[0, 0.5, 1]}
+    weight="bold"
+  >
+    <Text size={52}>􀙶</Text>
+  </GradientText>
+);
+
+const RestoreIcloudStep = ({ userData }) => {
+  const { goBack, replace } = useNavigation();
+  const dispatch = useDispatch();
+  const { isWalletLoading } = useWallets();
+  const initializeWallet = useInitializeWallet();
+  const { accountAddress } = useAccountSettings();
+  const [validPassword, setValidPassword] = useState(false);
+  const [incorrectPassword, setIncorrectPassword] = useState(false);
+  const [passwordFocused, setPasswordFocused] = useState(true);
+  const [password, setPassword] = useState('');
+  const [label, setLabel] = useState('􀎽 Confirm Backup');
+  const passwordRef = useRef();
+  const { setComponent, hide } = usePortal();
+
+  useEffect(() => {
+    if (isWalletLoading) {
+      setComponent(
+        <LoadingOverlay
+          paddingTop={sheetVerticalOffset}
+          title={isWalletLoading}
+        />,
+        false
+      );
+    }
+    return hide;
+  }, [hide, isWalletLoading, setComponent]);
+
+  useEffect(() => {
+    const fetchPasswordIfPossible = async () => {
+      const pwd = await fetchBackupPassword();
+      if (pwd) {
+        setPassword(pwd);
+      }
+    };
+    fetchPasswordIfPossible();
+  }, []);
+
+  const onPasswordFocus = useCallback(() => {
+    setPasswordFocused(true);
+  }, []);
+
+  const onPasswordBlur = useCallback(() => {
+    setPasswordFocused(false);
+  }, []);
+
+  useEffect(() => {
+    let newLabel = '';
+    let passwordIsValid = false;
+
+    if (incorrectPassword) {
+      newLabel = 'Incorrect Password';
+    } else {
+      if (isCloudBackupPasswordValid(password)) {
+        passwordIsValid = true;
+      }
+
+      newLabel = `􀑙 Restore from iCloud`;
+    }
+
+    setValidPassword(passwordIsValid);
+    setLabel(newLabel);
+  }, [incorrectPassword, password, passwordFocused]);
+
+  const onPasswordChange = useCallback(
+    ({ nativeEvent: { text: inputText } }) => {
+      setPassword(inputText);
+      setIncorrectPassword(false);
+    },
+    []
+  );
+
+  const onSubmit = useCallback(async () => {
+    try {
+      await dispatch(setIsWalletLoading(WalletLoadingStates.RESTORING_WALLET));
+      const success = await restoreCloudBackup(password, userData);
+      if (success) {
+        // Store it in the keychain in case it was missing
+        await saveBackupPassword(password);
+        // Set backup state
+        await saveUserBackupState(BackupStateTypes.done);
+        // Get rid of the current wallet
+        await removeWalletData(accountAddress);
+
+        // Reload the wallets
+        await dispatch(walletsLoadState());
+        await initializeWallet();
+        goBack();
+        InteractionManager.runAfterInteractions(async () => {
+          replace(Routes.SWIPE_LAYOUT);
+          await dispatch(setIsWalletLoading(null));
+        });
+      } else {
+        setIncorrectPassword(true);
```

Comment:
Should we alert the user here about the incorrect password failure

Reply:
We're already setting the button label as "Incorrect Password" when this flag is set._

Reply:
Ah yes